src/3d/qgs3dmapscene.cpp

```diff
@@ -247,6 +248,60 @@ void Qgs3DMapScene::viewZoomFull()
   mCameraController->resetView( 1.5 * std::sqrt( a * a - side * side ) );  // assuming FOV being 45 degrees
 }
 
+void Qgs3DMapScene::setViewFrom2DExtent( const QgsRectangle &extent )
+{
+  QgsPointXY center = extent.center();
+  QgsVector3D centerWorld = mMap.mapToWorldCoordinates( QVector3D( center.x(), center.y(), 0 ) );
+  QgsVector3D p1 = mMap.mapToWorldCoordinates( QVector3D( extent.xMinimum(), extent.yMinimum(), 0 ) );
+  QgsVector3D p2 = mMap.mapToWorldCoordinates( QVector3D( extent.xMaximum(), extent.yMaximum(), 0 ) );
+
+  float xSide = std::abs( p1.x() - p2.x() );
+  float ySide = std::abs( p1.z() - p2.z() );
+  if ( xSide < ySide )
+  {
+    float fov = 2 * std::atan( std::tan( qDegreesToRadians( cameraController()->camera()->fieldOfView() ) / 2 ) * cameraController()->camera()->aspectRatio() );
+    float r = xSide / 2.0f / std::tan( fov / 2.0f );
+    mCameraController->setViewFromTop( centerWorld.x(), centerWorld.z(), r );
+  }
+  else
+  {
+    float fov = qDegreesToRadians( cameraController()->camera()->fieldOfView() );
+    float r = ySide / 2.0f / std::tan( fov / 2.0f );
+    mCameraController->setViewFromTop( centerWorld.x(), centerWorld.z(), r );
+  }
+}
+
+QVector<QgsPointXY> Qgs3DMapScene::viewFrustum2DExtent()
+{
+  Qt3DRender::QCamera *camera = mCameraController->camera();
+  const QRect viewport = mCameraController->viewport();
+  QVector<QgsPointXY> extent;
+  QVector<int> pointsOrder = { 0, 1, 3, 2 };
+  for ( int i : pointsOrder )
+  {
+    const QPoint p( ( ( i >> 0 ) & 1 ) ? 0 : viewport.width(), ( ( i >> 1 ) & 1 ) ? 0 : viewport.height() );
+    QgsRay3D ray = Qgs3DUtils::rayFromScreenPoint( p, viewport.size(), camera );
+    QVector3D dir = ray.direction();
+    if ( dir.y() == 0.0 )
+      dir.setY( 0.000001 );
+    double t = - ray.origin().y() / dir.y();
+    if ( t < 0 )
+    {
+      // If the projected point is on the back of the camera we choose the farthest point in the front
+      t = camera->farPlane();
+    }
+    else
+    {
+      // If the projected point is on the front of the camera we choose the closest between it and farthest point in the front
+      t = std::min<float>( t, camera->farPlane() );
+    }
+    QVector3D planePoint = ray.origin() + t * dir;
```

Comment:
I think you need to separately examine the case where the camera is above the terrain but yet below the camera rotation point. In that case a rectangle is returned which is kind of unexpected.